added hardcallIterater to GenotypeStream

jbloom22 opened the pull request:
- added skipLEB128 to byte iterator
- modifed test in GenotypeStreamSuite

Before merging, we should run a larger number of generated examples through the test. @cseed any advise?

jbloom22 commented:
@cseed I changed count to use hardcallIterator, but the timing on my laptop on profile225.vds is about 14.5s with this branch and with master.  Could the overhead be elsewhere?  With genotypes=False it's 3.5s.

```
hc.read('/Users/jbloom/data/profile225.vds').count(genotypes=True)
```

cseed commented:
On the full chr22 I get a 54% speedup:

hardcall_iterator:
```
  count: 16.154s
  count: 16.584s
  count: 15.765s
```

master:
```
  count: 34.908s
  count: 35.148s
  count: 35.348s
```

Are you sure you built the right branch?  Rebuilt shadowJar if that's what you're using?

Anyway, looks good.  I'll have a look at the code.

cseed commented on src/main/scala/is/hail/variant/GenotypeStream.scala:
```diff
@@ -12,13 +12,17 @@ import is.hail.utils.ByteIterator
+  override def next(): Genotype = Genotype.read(nAlleles, isDosage, b)
+}
+
+// FIXME: Implement unboxed Int iterator
```
You want to do this?  I'm in favor.  I think it should look like this:

```
abstract class IntIterator extends Iterator[Int] {
  def nextInt(): Int
  override def next(): Int = nextInt()
}
```

and `def hasNext: Boolean` is still abstract.  That way it can be used without boxing by calling `nextInt` but it is still integrated with Scala collections.

jbloom22 commented:
Thanks, I added the IntIterator to Utils and switched hardcallIterator to extend it.  Back to you.

I definitely switched branches and used shadowJar (I went back and forth twice because I didn't believe the results).

cseed commented on src/main/scala/is/hail/driver/Count.scala:
```diff
@@ -31,8 +29,8 @@ object Count extends Command {
-      val (nVar, nCalled) = vds.rdd.map { case (v, (va, gs)) =>
-        (1L, gs.count(_.isCalled).toLong)
+      val (nVar, nCalled) = vds.rdd.map { case (v, (va, gs: GenotypeStream)) =>
+        (1L, gs.hardcallIterator.count(_ >= 0).toLong)
```
Use the unboxed interface here.

cseed commented:
I'll accept this as-is, but you can also use it where relevant.  Here's a list that springs to mind:

Places that immediately spring to mind: IBD, PCA (`ToStandardizedIndexRowMatrix`), regressions, Hardcalls (looking at that, we preserve fakeRef so that might not work), concordance (I think), export_plink (`ExportBedBimFam.makeBedRow`), GRM, tdt  (You'll need to define Genotype.getType for an unboxedGT.  Move it to the Genotype companion object.)

cseed commented on src/main/scala/is/hail/driver/Count.scala:
```diff
@@ -31,8 +29,8 @@ object Count extends Command {
     val (nVariants, nCalledOption) = if (options.genotypes) {
-      val (nVar, nCalled) = vds.rdd.map { case (v, (va, gs)) =>
-        (1L, gs.count(_.isCalled).toLong)
+      val (nVar, nCalled) = vds.rdd.map { case (v, (va, gs: GenotypeStream)) =>
```
The `Iterable[Genotype]` may not be a `GenotypeStream`, for example, if we ran `filter_genotypes.`  You should make a helper function somewhere that takes an `Iterable[Genotype]` and returns a `Iterable[Int]` and uses `hardcallIterator` if it is a `GenotypeStream` and `gs.iterator.map(_.unboxedGT)` otherwise (we'll, you'll have to create an IntIterator).

jbloom22 commented:
@cseed please take another look. I haven't had a chance to expand to other commands, but happy to do so later.

cseed commented on src/main/scala/is/hail/driver/Count.scala:
```diff
@@ -32,7 +30,7 @@ object Count extends Command {
     val (nVariants, nCalledOption) = if (options.genotypes) {
       val (nVar, nCalled) = vds.rdd.map { case (v, (va, gs)) =>
-        (1L, gs.count(_.isCalled).toLong)
+        (1L, ToIteratorGT(gs).count(_ >= 0).toLong)
```
Use the unboxed implementation here.

cseed commented on src/main/scala/is/hail/variant/GenotypeStream.scala:
```diff
@@ -116,6 +128,14 @@ object GenotypeStream {
 }
 
+object ToIteratorGT {
+  def apply(it: Iterable[Genotype]): Iterator[Int] =
```
This should return an `IntIterator`.

cseed replied:
It might be clearer to add an implicit from `Iterable[Genotype]` to `RichIterableGenotype` with a `hardcallIterator` member.  Then you can do `gs.hardcallIterator` for `gs: Iterable[Genotype]`.

jbloom22 replied:
I agree, I've moved it to implicit. And in order to make hardCallIterator return an IntIterator that is not a HardcallGenotypeStreamIterator, I created a HardCallIterator.  Back to you.

cseed commented on src/main/scala/is/hail/driver/Count.scala:
```diff
@@ -32,7 +29,7 @@ object Count extends Command {
     val (nVariants, nCalledOption) = if (options.genotypes) {
       val (nVar, nCalled) = vds.rdd.map { case (v, (va, gs)) =>
-        (1L, gs.count(_.isCalled).toLong)
+        (1L, gs.hardCallIterator.count(_ >= 0).toLong)
```
Make this use the unboxed interface!!

jbloom22 replied:
I clearly don't understand what you're asking for. hardCallIterator creates an HardCallGenotypeStreamIterator which is an IntIterator which is unboxed.

cseed replied:
Which implements `Iterator[Int]` which is boxed.  `Iterator[T].count()` calls `Iterator[T].next(): T` which causes the Int to be boxed.

jbloom22 replied:
I added count to IntIterator:

```
  override def count(p: Int => Boolean): Int = {
    var cnt = 0
    while (hasNext) {
      if (p(nextInt())) cnt += 1
    }
    cnt
  }
```

cseed commented on src/main/scala/is/hail/utils/richUtils/RichIterableGenotype.scala:
```diff
@@ -0,0 +1,8 @@
+import is.hail.variant.Genotype
+
+class RichIterableGenotype(it: Iterable[Genotype]) {
+  def hardCallIterator: HardCallIterator = new HardCallIterator(it.iterator)
```
With this definition, you never use your new code.  You have to match on the type of it here.

I'd probably define `HardCallIterator` inline since it is (should be) only used once.  This should return an `IntInterator`.

The goal here (as with the `MutableGenotype` stuff) is eliminate all per-Genotype allocations.

jbloom22 replied:
I thought the implicit would only be called if gs was not a GenotypeStream, since that's the more specialized class on which hardCallIterator is defined.

Is this what you have in mind?
```
class RichIterableGenotype(it: Iterable[Genotype]) {
  def hardCallIterator: IntIterator = it match {
    case gs: GenotypeStream => gs.hardCallIterator
    case _ =>
      val iter = it.iterator
      class HardCallIterator extends IntIterator {
        override def hasNext: Boolean = iter.hasNext
        override def nextInt(): Int = iter.next().unboxedGT
      }
      new HardCallIterator
  }
}
```

Or this?

```
class RichIterableGenotype(it: Iterable[Genotype]) {
  def hardCallIterator: IntIterator = it match {
    case gs: GenotypeStream => gs.hardCallIterator
    case _ =>
      val iter = it.map(_.unboxedGT).iterator
      class HardCallIterator extends IntIterator {
        override def hasNext: Boolean = iter.hasNext
        override def nextInt(): Int = iter.next()
      }
      new HardCallIterator
  }
}
```

 Or something else entirely?

cseed replied:
You can't dynamically choose to resolve through an implicit vs a virtual function.  In particular, Iterable[Int] doesn't have a hardCallIterator method so it will call it, even if it is pointing to a subclass that does.

Reusing the same name does could cause confusion.  You might rename the method on GenotypeStream.

I think we normally reserve `it` for `Iterator`s.

You can instantiate a subclass without naming it.

```
class RichIterableGenotype(gs: Iterable[Genotype]) {
  def hardCallIterator: IntIterator = gs match {
    case gs: GenotypeStream => gs.hardCallIterator
    case _ =>
      new IntIterator {
        val it = gs.iterator
        override def hasNext: Boolean = it.hasNext
        override def nextInt(): Int = it.next().unboxedGT
      }
  }
}
```

jbloom22 replied:
I renamed to gsHardCallIterator on Genotype Stream.

cseed commented on src/main/scala/is/hail/utils/IntIterator.scala:
```diff
@@ -0,0 +1,14 @@
+  override def next(): Int = nextInt()
+}
+
+class HardCallIterator(val it: Iterator[Genotype]) extends IntIterator {
```
This shouldn't be in here.  Or anywhere, see my comment below.

jbloom22 replied:
I removed it.

cseed commented on src/main/scala/is/hail/utils/IntIterator.scala:
```diff
@@ -0,0 +1,14 @@
+  def nextInt(): Int
+  override def next(): Int = nextInt()
+
+  override def count(p: Int => Boolean): Int = {
```
`Int => Boolean` is a `Function1[Int, Boolean]`.  That means you allocate a Function1 object (per-variant, OK), but the both the Int and Boolean are boxed completely obviating the benefit of `IntIterator`.

In this case that isn't true since Function1 is specialized for in the Int => Boolean case, but not for all combinations of primitive types.  The specialization isn't even documented:

http://www.scala-lang.org/api/current/scala/Function1.html

you need to look at the code:

https://github.com/scala/legacy-svn-scala/blob/master/src/library/scala/Function1.scala#L29

so I think this sets a bad precedent.  My instinct is now "if it looks like a Scala collection it performs badly" and should not be used in per-genotype operations.

jbloom22 replied:
Ok, that makes sense, and I read more about type erasure and specialization:
http://www.scala-notes.org/2011/04/specializing-for-primitive-types/

I've hard-coded the function as follows:

```
  def countNonNegative(): Int = {
    var cnt = 0
    while (hasNext) {
      if (nextInt() >= 0) cnt += 1
    }
    cnt
  }
```
